Adds template for Cloud Native Buildpacks

ekcasey opened the pull request:
Intended to replace https://github.com/knative/build-templates/pull/65

* We made some changes to address the comments on the above PR
* We also rearranged the lifecycle components into separate images and added a `knative-helper`

jonjohnsonjr commented on buildpacks/cnb.yaml:
```diff
@@ -0,0 +1,54 @@
+
+  steps:
+  - name: prepare
+    image: packs/base
```
I don't love all of these being `:latest` and `imagePullPolicy: Always`. That gives the `packs` owners the ability to break this BuildTemplate inadvertently by pushing new images. WDYT about versioning these by digest? There's also a small perf win if you use digests + `imagePullPolicy: IfNotPresent` in not having to ping dockerhub to see if the image has changed.

As of now:

```
packs/base@sha256:0d6949aaa62649223ca13c8a02a03584068e1e5bdd8ed6edd992e107603b083c
packs/build@sha256:99bcaa9904d7eef78df9c04ddf22dc9d78b0d06d548c68c03a71d04704df3048
packs/util@sha256:f97205c746358d03b4aa84480335f04cd69b25a0348f85a6a5a1c7e9aa7f2a86
```

OTOH, that means you can't fix things in the wild, but I tend to favor reproducibility /shrug

mattmoor replied:
FWIW, Digests will work better with the caching model I'm adding to BuildTemplates and provide a much better story for release signoff on a particular shape of buildtemplate.

ekcasey replied:
We could use digests for `packs/util` because it contains utilities that won't change often, but it is more problematic for `packs/build`. 

Part of the value prop of buildpacks is that they keep your application and system dependencies up to date to avoid CVEs. A buildpack version comes with a tested set of dependencies (for an example look at http://buildpacks.cloudfoundry.org/#/buildpacks, these are CF buildpacks but the Cloud Native Buildpacks will have the same dependency model). New buildpack versions are released regularly with updated dependencies. When new buildpack versions are released we need to bump the buildpacks in the `packs/build` image. 

This means we will be releasing `packs/build` frequently. If we pinned the image to a digest we would need to make an awful lot of PRs, and users would need to constantly reapply the template to get desired security benefits.

sclevine replied:
This is also intrinsic to the CNB model -- we build directly on the build image so that we don't have to nest containers. That means any dependencies that are statically linked into the app binaries won't get CVE updates until you re-build against the latest build image.

Eventually, we may consider using tags to provide buildpack/OS-package updates without changing the CNB code itself. But for now, we'd like each build to use the latest build image.

jonjohnsonjr replied:
> Part of the value prop of buildpacks is that they keep your application and system dependencies up to date to avoid CVEs.

Ah, this makes lots of sense, ignore me! 👍

ekcasey commented:
I was just added to the Pivotal group of authorized contributors so I am covered by the corporate CLA.

ImJasonH commented on buildpacks/cf.yaml:
```diff
@@ -1,7 +1,7 @@
 kind: BuildTemplate
 metadata:
-  name: buildpack
+  name: cloudfoundry
```
Should we just remove this build template? Or maybe rename it to `cloudfoundry-buildpack`?

sclevine replied:
The CNB builders don't support the Cloud Foundry buildpacks yet.

ImJasonH commented on buildpacks/README.md:
```diff
@@ -0,0 +1,44 @@
+      url: https://github.com/my-user/my-repo
+      revision: master
+  template:
+    name: cnb
```
`cnb` is not very descriptive, how about just `buildpack` ?

sclevine replied:
Maybe `buildpack-cnb` for this, and `buildpack-cf` for the CF one?

ImJasonH replied:
I think the acronyms `cnb` and `cf` might be confusing for people who are not already experts in the area.

I'd like to get to the point (maybe not in the PR, but soonish) where we speak in terms of `buildpack` and `buildpack-old` (or `buildpack-cloudfoundry` or something else), so users who come here saying "I'd like to use buildpacks" know which one is for them.

sclevine replied:
How about `buildpack-cnb` and `buildpack-cloudfoundry` -- just for now?

These templates are targeted at different users:
- The Cloud Foundry template provides production-ready CF buildpacks for app devs
- The CNB template is for buildpack creators who want to create/run CNB-compatible buildpacks on knative.

I'd like to get to a point where both templates are CNB templates, but one provides Cloud Foundry buildpacks and the other provides riff function invoker buildpacks. (@jldec)

ImJasonH commented on buildpacks/cnb.yaml:
```diff
@@ -0,0 +1,54 @@
+metadata:
+  name: cnb
+spec:
+  parameters:
```
The old buildpack template supported a `CACHE` parameter, does buildpacks v3 support that today? It would be great if we could just change buildpack.yaml to run the new v3 images under the hood, without changing parameters at all.

sclevine replied:
Definitely agree that we should keep this parameter so that an interbuild cache volume can be used.

There are two reasons we have separate templates with separate arguments for the CF and CNB models:
- The models are very different. CNB:
  - requires a separate analyze step to pull metadata from the previous image
  - requires detection
  - requires a more complex buildpack order definition
- The included buildpacks are different and incompatible. The CF template uses the Cloud Foundry buildpacks (by retrofitting the Cloud Foundry lifecycle into builder images). The CNB template currently includes [these samples](https://github.com/buildpack/samples) and may eventually include function invoker buildpacks.

ericbottard commented on buildpacks/README.md:
```diff
@@ -0,0 +1,44 @@
+> image format. They take advantage of the latest capabilities of these standards, such as remote
+> image layer rebasing on Docker API v2 registries.
+
+**Note**: The current Cloud Foundry buildpacks are available in the [CF template](README-CF.md).
```
This may seem obvious to you, but what about clearly stating that "the current CF buildpacks" are *not* v3 buildpacks, which we're dealing with here?

ekcasey replied:
We could explicitly call this out now. However, v3 compatible CF buildpacks are being developed and will be released in coming months.

ericbottard commented on buildpacks/cnb.yaml:
```diff
@@ -0,0 +1,60 @@
+      mountPath: /cache
+    imagePullPolicy: Always
+  - name: detect
+    image: packs/samples
```
How about parameterizing this value, thus allowing users to specify the buildpack (or rather "buildpack-builder") they want to use? This parameter would apply to line 41, of course

ekcasey replied:
We can't do this until `knative-build` 0.2 is released. The currently released version doesn't support parameterizing images.

ekcasey commented:
@ImJasonH Where are we on this PR? Are any of these concerns still outstanding?

jchesterpivotal commented:
/retest

jchesterpivotal commented:
Looking at the e2e test failure, it seems as though [we broke it by renaming `buildpack/buildpack.yaml` to `buildpack/cf.yaml`](https://github.com/knative/build-templates/blob/master/test/e2e-tests.sh#L33).

@ekcasey I think we'll need to edit this PR to get it through the e2e spec.

ekcasey commented:
@jchesterpivotal and I fixed the e2e tests and added a new one for the `buildpacks-cnb` template.

@ImJasonH @jonjohnsonjr let us know if you have any comments :)

shashwathi commented:
/cla check

shashwathi commented:
/check-cla

shashwathi commented:
/assign @ImJasonH

ekcasey commented:
We made some significant changes to this PR, so that the build template uses our the new `v3alpha2` lifecycle. Now that the step image is configurable we also added some params to allow users to supply their own builder image.

We also fixed the template names in the buildpacks READMEs.

matthewmcnew commented:
I was just added to the Pivotal group of authorized contributors so I should be covered by the corporate CLA. 
I am ok with my commits being contributed to this project

shashwathi commented:
/check-cla

ekcasey commented on buildpacks/README.md:
```diff
@@ -20,24 +20,28 @@ kubectl apply -f https://raw.githubusercontent.com/knative/build-templates/maste
-* **IMAGE:** The Docker image name to apply to the newly built image. (_required_)
+* **IMAGE:** The image you wish to create. For example, "repo/example", or "example.com/repo/image". (_required_)
 * **RUN_IMAGE:** The run image buildpacks will use as the base for IMAGE. (_default:_ `packs/run`)
+* **BUILDER_IMAGE** The image on which builds will run. ( default: `packs/samples:packs/samples:v3alpha2`)
```
good catch @jchesterpivotal. just fixed it

jamesward commented:
Should the exporter step mount the docker socket and use `-daemon`?

sclevine commented:
The CNB exporter is much less efficient when exporting to a daemon instead of a registry. When exporting to a registry, the exporter can upload just the layers that have changed, regardless of their order within the image.

Additionally, the whole process currently works in unprivileged containers. Poking the daemon socket through would make it less suitable for building untrusted code. (That said, I'm not sure what privileges containers currently have in build CRD.)

ImJasonH commented:
/lgtm
